arch/x86/entry.S

```diff
@@ -32,38 +32,46 @@
 #include <usermode.h>
 #include <errno.h>
 
-.macro _handle_usermode cr3
+.macro _from_usermode switch_stack=1
+    SET_CR3 cr3
+    swapgs
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+.endm
+
+.macro _to_usermode switch_stack=1
+    .if \switch_stack == 1
+        SWITCH_STACK
+    .endif
+    swapgs
+    SET_CR3 user_cr3
+.endm
+
+.macro exc_from_usermode
     testb $0x3, cpu_exc_cs(%_ASM_SP)
     jz 1f /* skip if from kernel mode */
-        SET_CR3 \cr3
-        swapgs
+        _from_usermode switch_stack=0
     1:
 .endm
```

Comment:
hm this reads a bit funny, because we're not necessarily coming from usermode. it's more like `maybe_exc_from_usermode`... Does it make sense to move the test/jz out from this?

Reply:
Yes, I think we can wrap it up into a more explicit abstraction.